### DISCUSSION TITLE
v7 Beta: Use of RegionsPlugin is odd?

### DISCUSSION THREAD

**Discussion by jgentes** [katspaugh/wavesurfer.js#2791] [General]
Quick question - I was using Markers in v6 and now switching to Regions with v7.

I load the plugin with
```js
import RegionsPlugin from "wavesurfer.js/dist/plugins/regions.js";
```

And I initialize the plugin within WaveSurferOptions with
```js
RegionsPlugin.create()
```

But I'm surprised there's no easy way to access it for `addRegion` or `clearRegions` - is this correct?
```ts
const regions = wavesufer.getActivePlugins().find((plugin: RegionsPlugin) => plugin.regions);
regions.clearRegions();
```

This just seems awkward.. I know I could use the index instead (`wavesufer.getActivePlugins()[0]`) but that smells bad too, since the order of the plugins might change.

**Comment by katspaugh** [comment]
The intended way is

```
const regions = Regions.create()
const wavesurfer = WaveSurfer.create({
  plugins: [ regions ]
})

// alternatively

const regions = wavesurfer.registerPlugin(Regions.create())
```

I’m open to suggestions though.

**Comment by jgentes** [comment]
Thanks, I've been trying to think of another way to do it. I liked how the Markers plugin generated `wavesurfer.markers` which made it easily accessible, but I could see why you don't want to pollute the API surface with unknown plugin accessors.

I don't really want to pass `regions` around in my code since it's available on `wavesurfer`. I just have to use `getActivePlugins()` to reach it.

One idea could be to expand `getActivePlugins` to allow for a name, since they do have defined names like `RegionsPlugin`, so the code could be:
```js
const regions = wavesurfer.getActivePlugins('RegionsPlugin');
```

**Reply by katspaugh** [reply]
The problem is the plugins array is… an array, and there can be multiple plugins with the same name. E.g. see the Timeline x2 example with two timelines with different settings.

**Comment by treio** [comment]
You can easily extend the original class...  

For example, the following is what I have:

```typescript
import WaveSurfer, { type WaveSurferOptions } from "wavesurfer.js";
import SubtitlesPlugin from "./SubtitlesPlugin";
import TimelinePlugin from "wavesurfer.js/dist/plugins/timeline";
import MinimapPlugin from "wavesurfer.js/dist/plugins/minimap";
import { type SubtitleObject } from "~/routes/_app.sync";

/**
 * This class is a subclass of WaveSurfer that automatically adds and registers plugins,
 * and provides some other methods that may be handful.
 *
 * Original repo:
 * https://github.com/katspaugh/wavesurfer.js
 *
 * Homepage:
 * https://wavesurfer.pages.dev/
 */
class EnhancedWavesurfer extends WaveSurfer {
  private _subtitlesPlugin: SubtitlesPlugin;
  private _timelinePlugin: TimelinePlugin;
  private _minimapPlugin: MinimapPlugin;

  /**
   * The function creates and returns a new instance of
   * the EnhancedWavesurfer class with the provided options.
   */
  static create(options: WaveSurferOptions): EnhancedWavesurfer {
    return new EnhancedWavesurfer(options);
  }

  /**
   * This is a constructor function that creates a new instance of
   * a WaveSurfer object with already registered plugins.
   */
  constructor(options: WaveSurferOptions) {
    super({
      // Some pre-defined values
      height: 150,
      waveColor: "rgb(25, 118, 210)",
      progressColor: "rgba(255, 255, 255, 0.25)",
      cursorColor: "rgb(255, 0, 0)",
      minPxPerSec: 60,
      autoCenter: false,
      barWidth: 1,
      //barGap: 1,
      //barRadius: 1,

      // Other values
      ...options,
    });

    // Automatically create and register plugins on initialization.

    this._subtitlesPlugin = new SubtitlesPlugin();
    this.registerPlugin(this._subtitlesPlugin);

    this._timelinePlugin = new TimelinePlugin({ height: 20, insertPosition: "beforebegin" });
    this.registerPlugin(this._timelinePlugin);

    this._minimapPlugin = new MinimapPlugin({
      height: 20,
      waveColor: "#ddd",
      progressColor: "#999",
      container: options.container, // ? this.getWrapper(),
    });
    this.registerPlugin(this._minimapPlugin);
  }

  /**
   * This function returns the registered SubtitlesPlugin.
   */
  get subtitlesPlugin(): SubtitlesPlugin {
    return this._subtitlesPlugin;
  }
}

export default EnhancedWavesurfer;
```

Further, if you create `wavesurfer` instance using `EnhancedWavesurfer` instead, you can simply call `wavesurfer.subtitlesPlugin.on(...)` etc., and you are free to customize it as needed.

Hope that helps!

**Reply by jgentes** [reply]
Thanks @treio that's a nice approach, but I try to avoid recasting classes unless absolutely necessary. I've decided to go with the index approach: `wavesurfer.getRegions()[0]` and just be careful not to change the order in which they are instantiated, which isn't a good solution either :(
